rust/kernel/fs.rs

```diff
@@ -0,0 +1,1262 @@
+
+//! A kernel file system.
+//!
+//! This module allows Rust code to register new kernel file systems.
+//!
+//! C headers: [`include/linux/fs.h`](../../include/linux/fs.h)
+
+pub mod buffer;
+
+/// Maximum size of an inode.
+pub const MAX_LFS_FILESIZE: i64 = bindings::MAX_LFS_FILESIZE;
+
+/// Contains constants related to Linux file modes.
+pub mod mode {
+    /// A bitmask used to the file type from a mode value.
+    pub const S_IFMT: u32 = bindings::S_IFMT;
+
+    /// File type constant for block devices.
+    pub const S_IFBLK: u32 = bindings::S_IFBLK;
+
+    /// File type constant for char devices.
+    pub const S_IFCHR: u32 = bindings::S_IFCHR;
+
+    /// File type constant for directories.
+    pub const S_IFDIR: u32 = bindings::S_IFDIR;
+
+    /// File type constant for pipes.
+    pub const S_IFIFO: u32 = bindings::S_IFIFO;
+
+    /// File type constant for symbolic links.
+    pub const S_IFLNK: u32 = bindings::S_IFLNK;
+
+    /// File type constant for regular files.
+    pub const S_IFREG: u32 = bindings::S_IFREG;
+
+    /// File type constant for sockets.
+    pub const S_IFSOCK: u32 = bindings::S_IFSOCK;
+}
+
+/// Read-only file systems.
+pub mod ro {
+    use crate::error::{code::*, from_result, to_result, Error, Result};
+    use crate::types::{ARef, AlwaysRefCounted, Either, ForeignOwnable, Opaque};
+    use crate::{
+        bindings, container_of, folio::LockedFolio, init::PinInit, mem_cache::MemCache, str::CStr,
+        time::Time, try_pin_init, ThisModule,
+    };
+    use core::mem::{size_of, ManuallyDrop, MaybeUninit};
+    use core::{marker::PhantomData, marker::PhantomPinned, pin::Pin, ptr};
+    use macros::{pin_data, pinned_drop};
+
+    /// Type of superblock keying.
+    ///
+    /// It determines how C's `fs_context_operations::get_tree` is implemented.
+    pub enum Super {
+        /// Multiple independent superblocks may exist.
+        Independent,
+
+        /// Uses a block device.
+        BlockDev,
+    }
+
+    /// A read-only file system type.
+    pub trait Type {
+        /// Data associated with each file system instance (super-block).
+        type Data: ForeignOwnable + Send + Sync;
+
+        /// Type of data associated with each inode.
+        type INodeData: Send + Sync;
+
+        /// The name of the file system type.
+        const NAME: &'static CStr;
+
+        /// Determines how superblocks for this file system type are keyed.
+        const SUPER_TYPE: Super = Super::Independent;
+
+        /// Initialises a super block for this file system type.
+        fn fill_super(sb: NewSuperBlock<'_, Self>) -> Result<&SuperBlock<Self>>;
+
+        /// Reads directory entries from directory inodes.
+        ///
+        /// `pos` is the current position of the directory reader.
+        ///
+        /// `report` is a closure that takes the following arguments, in order: the entry name, its
+        /// position, its inode number, and its type. It returns `true` if the entry was copied to
```

Comment:
I wonder if there is a better way to document this sort of `Fn*`s signatures. Perhaps a `struct` with named parameters? But that means having to look up yet another type...

In any case, relatedly, would it make sense to have at a type alias for inode numbers (`Ino`?) or perhaps even a new type?

Reply:
I removed the closure in favour of an emitter object. This simplified things because it allowed me to push some of the logic of not failing when at least one entry was reported to core code.

I also added a type for inode numbers.